Stats reparent shard operation timings

timvaillancourt opened the pull request:
## Description

This PR improves reparenting stats on `vtctld` by:
1. Adding the same counters from ERS to PRS:
    - `prs_counter`
    - `prs_failure_counter`
    - `prs_success_counter`
2. Adding timing stats to both ERS and PRS:
    - `ers_timings`
    - `prs_timings`
3. Making all reparenting metrics include `Keyspace` as a label

## Related Issue(s)

Issue TBD

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

## Deployment Notes

<!-- Notes regarding deployment of the contained body of work. These should note any db migrations, etc. -->

deepthi commented:
We might as well go ahead and rename the metrics.
prs -> planned_reparent / PlannedReparent
ers -> emergency_reparent / EmergencyReparent

We'll have to keep the old ers metrics around for 1 release.

For metrics / stats PRs in general, it is helpful to paste the output from `/debug/vars` or `/debug/metrics` into the PR description so that reviewers can see what the users/consumers of the metrics would see.

It will be nice to document the new metrics on the website. Though I'm not sure the existing ones have been documented either.

timvaillancourt commented:
> We might as well go ahead and rename the metrics. prs -> planned_reparent / PlannedReparent ers -> emergency_reparent / EmergencyReparent

@deepthi great, I'll do that 👍
 
> We'll have to keep the old `ers` metrics around for 1 release, so we will need to have two sets of those for now.

@deepthi makes sense 👍 

> For metrics / stats PRs in general, it is helpful to paste the output from `/debug/vars` or `/debug/metrics` into the PR description so that reviewers can see what the users/consumers of the metrics would see.

👍 

> It will be nice to document the new metrics on the website. Though I'm not sure the existing ones have been documented either.

Sure, I can create a website PR for this 👍

Finally, I'd like to add similar metrics to `vtorc` if they don't exist already, and keeping metric naming/etc consistent between these two failover components could be useful. cc'ing @GuptaManan100 for some feedback. I can make a `vtorc` follow-up PR if needed

GuptaManan100 commented:
@timvaillancourt VTOrc uses the same code for running ERS and PRS like vtctld. I don't think that any other changes will be required to get these metrics in VTOrc. If you would be so kind to verify this, I would be much obliged.

timvaillancourt commented:
> @timvaillancourt VTOrc uses the same code for running ERS and PRS like vtctld. I don't think that any other changes will be required to get these metrics in VTOrc. If you would be so kind to verify this, I would be much obliged.

Thanks @GuptaManan100, that would be awesome. I'll confirm this 👍

One last question before I adjust this PR: would it make sense to have timings per "phase" of the reparent 🤔? Right now the timings added are for the whole operation, which has two drawbacks: if one subtask is slow it's not clear which, also it's not clear what the client impact was, because only a portion of the reparent has impact

For example, in PRS `.reparentShardLocked(...)` there are a few potential "phases" we could get timings for, but I'm not sure which is useful. There isn't a concept of "phases" in PRS:
1. `GetShard`
2. `GetKeyspaceDurability`
3. `GetTabletMapForShard`
4. `pr.verifyAllTabletsReachable`
5. `pr.preflightChecks`
6. `FindCurrentPrimary`
7. A few possible promotions:
    - `pr.performInitialPromotion`
    - `pr.performPotentialPromotion`
    - `pr.performPartialPromotionRecovery`
    - `pr.performGracefulPromotion`
8. `CheckShardLocked`
9. `pr.reparentTablets`
10. `pr.tmc.RefreshState`

Here is a wild guess of what "phases" we could track for PRS:
- `Init`(?) - this would cover steps 1 to 4 above
- `Preflight` - this would cover steps 5 above
- `Promotion` - this would cover step 7 above
- `Reparent` - this would cover step 9 above
- Not covered is `CheckShardLocked` and `prs.tmc.RefreshState`. Any thoughts there?

For ERS some phase ideas are:
- `Init`(?) - for [this block of ERS](https://github.com/timvaillancourt/vitess/blob/main/go/vt/vtctl/reparentutil/emergency_reparenter.go#L144-L195)
- `FindCandidate` - for [this block or ERS](https://github.com/timvaillancourt/vitess/blob/main/go/vt/vtctl/reparentutil/emergency_reparenter.go#L197-L260)
- `Promote` - for [this block](https://github.com/timvaillancourt/vitess/blob/main/go/vt/vtctl/reparentutil/emergency_reparenter.go#L262-L304) and [this block of ERS](https://github.com/timvaillancourt/vitess/blob/main/go/vt/vtctl/reparentutil/emergency_reparenter.go#L696-L716)
- `Reparent` - for [this block of ERS](https://github.com/timvaillancourt/vitess/blob/main/go/vt/vtctl/reparentutil/emergency_reparenter.go#L717-L723)

I'm curious if this is a good idea or overkill, also if the "phases" I'm proposing make sense. cc @deepthi / @GuptaManan100 for thoughts

GuptaManan100 commented:
I don't think timing for various components is super important. If a PRS/ERS fails, we can always figure out how long each step took based on the logs.

timvaillancourt commented:
> I don't think timing for various parts is super important. If a PRS/ERS fails, we can always figure out how long each step took based on the logs.

@GuptaManan100 sounds good 👍. Let's start with the changes discussed above. I'll update the PR

timvaillancourt commented:
Thanks for the feedback @GuptaManan100. Sounds good 👍

I've updated the PR with the suggestions @deepthi provided [in this comment](https://github.com/vitessio/vitess/pull/13723#issuecomment-1673812452)

Please let me know what you think about the new naming 🙇

GuptaManan100 commented on go/vt/vtctl/reparentutil/emergency_reparenter.go:
```diff
@@ -70,10 +70,20 @@ type EmergencyReparentOptions struct {
-	ersCounter        = stats.NewGauge("ers_counter", "Number of times Emergency Reparent Shard has been run")
-	ersSuccessCounter = stats.NewGauge("ers_success_counter", "Number of times Emergency Reparent Shard has succeeded")
-	ersFailureCounter = stats.NewGauge("ers_failure_counter", "Number of times Emergency Reparent Shard has failed")
+	ersCounter = stats.NewCountersWithMultiNamesAndLabels([]string{"ers_counter", "emergency_shard_reparents"},
+		"Number of times Emergency Reparent Shard has been run",
```
For backward compatibility, we shouldn't change the type of `ers_counter` from just a gauge to a counter that has labels associated. 
If a user has some kind of parsing setup on this metric, we don't want to break their flow. So I think we should leave all the old metrics just the way they are and introduce the new ones.

timvaillancourt replied:
@GuptaManan100 makes sense 👍. Updated to preserve the old stats as is

GuptaManan100 reviewed (commented):
Changes in the release notes are also required. We should mention the new metrics there and that the old metrics are being deprecated.

timvaillancourt commented:
> @timvaillancourt VTOrc uses the same code for running ERS and PRS like vtctld. I don't think that any other changes will be required to get these metrics in VTOrc. If you would be so kind to verify this, I would be much obliged.

Confirming that `vtorc` uses the same code [here](https://github.com/vitessio/vitess/blob/main/go/vt/vtorc/logic/topology_recovery.go#L233-L255) and [here](https://github.com/vitessio/vitess/blob/main/go/vt/vtorc/logic/topology_recovery.go#L788-L805) 🎉 

> Changes in the release notes are also required. We should mention the new metrics there and that the old metrics are being deprecated.

Sure, adding that 👍

timvaillancourt commented on go/vt/vtctl/reparentutil/emergency_reparenter.go:
```diff
@@ -71,9 +71,20 @@ type EmergencyReparentOptions struct {
+	ersFailureCounter = stats.NewCountersWithMultiLabels("emergency_shard_reparents_failed", "Number of times Emergency Reparent Shard has failed",
+		[]string{"Keyspace", "Shard"},
+	)
+	ersSuccessCounter = stats.NewCountersWithMultiLabels("emergency_shard_reparents_succeeded", "Number of times Emergency Reparent Shard has succeeded",
```
`emergency_shard_reparents_succeess` is another option 🤔

timvaillancourt commented on go/vt/vtctl/reparentutil/planned_reparenter.go:
```diff
@@ -41,6 +42,19 @@ import (
+	prsFailureCounter = stats.NewCountersWithMultiLabels("planned_shard_reparents_failed", "Number of times Planned Reparent Shard has failed",
+		[]string{"Keyspace", "Shard"},
+	)
+	prsSuccessCounter = stats.NewCountersWithMultiLabels("planned_shard_reparents_succeeded", "Number of times Planned Reparent Shard has succeeded",
```
`planned_shard_reparents_succeess` is another option here, too

timvaillancourt commented on go/vt/vtctl/reparentutil/util.go:
```diff
@@ -41,6 +42,8 @@ import (
 	"vitess.io/vitess/go/vt/proto/vtrpc"
 )
 
+var reparentShardTimings = stats.NewTimings("reparent_shard_timings", "Timings of reparent shard operations", "Operation")
```
`reparent_shard_operation_timings` is another option here. Any opinions?

timvaillancourt replied:
Made this change on second thought 👍

timvaillancourt commented on go/vt/vtctl/reparentutil/emergency_reparenter.go:
```diff
@@ -101,26 +112,33 @@ func NewEmergencyReparenter(ts *topo.Server, tmc tmclient.TabletManagerClient, l
+	startTime := time.Now()
 	ev := &events.Reparent{}
 	defer func() {
+		reparentShardOpTimings.Add("EmergencyReparentShard", time.Since(startTime))
```
Should we track the timings of all operations or success only? 🤔

deepthi replied:
May as well do all of them

timvaillancourt commented on go/vt/vtctl/reparentutil/planned_reparenter.go:
```diff
@@ -103,18 +119,23 @@ func (pr *PlannedReparenter) ReparentShard(ctx context.Context, keyspace string,
+	startTime := time.Now()
 	ev := &events.Reparent{}
 	defer func() {
+		reparentShardOpTimings.Add("PlannedReparentShard", time.Since(startTime))
```
Repeating comment from ERS: should we track the timings of all operations or success only? 🤔

deepthi replied:
Same comment as for ERS. Let's do all.

timvaillancourt commented:
@GuptaManan100 / @deepthi re-requesting review as I believe I've addressed the PR suggestions

Here are the added changelog sections:
1. https://github.com/vitessio/vitess/blob/3edfaf46bbb86551de92223d6e9c57a664881087/changelog/18.0/18.0.0/summary.md#deprecated-stats 
2. https://github.com/vitessio/vitess/blob/3edfaf46bbb86551de92223d6e9c57a664881087/changelog/18.0/18.0.0/summary.md#vtctld-and-vtorc-reparenting-stats-improvements

Changelog nits appreciated 🙇

GuptaManan100 commented on go/vt/vtctl/reparentutil/planned_reparenter.go:
```diff
@@ -495,6 +516,8 @@ func (pr *PlannedReparenter) reparentShardLocked(
 	shard string,
 	opts PlannedReparentOptions,
 ) error {
+	prsCounter.Add([]string{keyspace, shard}, 1)
+
```
Shouldn't we be incrementing `prsCounter` when we call `ReparentShard` instead of `reparentShardLocked`?

I am thinking that `prsCounter` should be equal to `prsSuccessCounter + prsFailureCounter`.

timvaillancourt replied:
@GuptaManan100 that makes sense. I've updated the new counters to increment in `ReparentShard` 👍

I left the legacy ERS counter in `reparentShardLocked` to avoid changing behaviour for that legacy stat

GuptaManan100 commented on go/vt/vtctl/reparentutil/emergency_reparenter.go:
```diff
@@ -144,7 +162,8 @@ func (erp *EmergencyReparenter) getLockAction(newPrimaryAlias *topodatapb.Tablet
 	// log the starting of the operation and increment the counter
 	erp.logger.Infof("will initiate emergency reparent shard in keyspace - %s, shard - %s", keyspace, shard)
-	ersCounter.Add(1)
+	legacyERSCounter.Add(1)
+	ersCounter.Add([]string{keyspace, shard}, 1)
```
Same comment as PRS

timvaillancourt replied:
Updated 👍

GuptaManan100 reviewed (commented):
If we are incrementing `prs_counter` in `reparentShardLocked` then we shouldn't be treating not being able to acquire a shard lock as an error, otherwise we would end up in the state, where we report more failures and successes than the number of times prs has run

Rest looks good to me!

GuptaManan100 commented on go/vt/vtctl/reparentutil/emergency_reparenter.go:
```diff
@@ -101,26 +112,34 @@ func NewEmergencyReparenter(ts *topo.Server, tmc tmclient.TabletManagerClient, l
 	if err = topo.CheckShardLocked(ctx, keyspace, shard); err != nil {
 		var unlock func(*error)
 		ctx, unlock, err = erp.ts.LockShard(ctx, keyspace, shard, opts.lockAction)
 		if err != nil {
+			legacyERSFailureCounter.Add(1)
+			ersFailureCounter.Add(statsLabels, 1)
 			return nil, err
 		}
```
This is still incorrect...
We are incrementing the failure counter and returning from the function without increasing the total count. So the total count would still not be the sum of successes and failures.
Also we shouldn't be changing the behaviour of the legacy counters.

timvaillancourt replied:
@GuptaManan100 makes sense, updated 👍

As for the `DCO` CI failure, it seems the rebase suggested by the failed check is getting rejected for reasons I cannot explain:
```
tvaillancourt@tvailla-ltmxctu vitess % git push --force-with-lease origin stats-reparent_shard_operation_timings
To https://github.com/timvaillancourt/vitess
 ! [rejected]              stats-reparent_shard_operation_timings -> stats-reparent_shard_operation_timings (stale info)
error: failed to push some refs to 'https://github.com/timvaillancourt/vitess'
```
I'm curious if you had ideas here 🤔

GuptaManan100 replied:
I have no idea! Maybe @deepthi knows?

timvaillancourt replied:
@GuptaManan100 I managed to resolve it by squashing my commits + signing 🎉. Thanks for checking!

GuptaManan100 reviewed (commented):
Same comment for PRS too

GuptaManan100 reviewed (approved):
LGTM now!
We require website updates for the PR too, but otherwise it looks good!

timvaillancourt commented:
> LGTM now! We require website updates for the PR too, but otherwise it looks good!

Thanks @GuptaManan100. Where in the website docs should this be added?

I see [there are docs for VReplication metrics](https://vitess.io/docs/16.0/reference/vreplication/metrics/) and [general monitoring docs](https://vitess.io/docs/17.0/user-guides/configuration-basic/monitoring/), but I couldn't spot a natural place to put docs for this. Maybe there is no existing doc for `vtctld`/`vtorc` metrics?

GuptaManan100 commented:
VTOrc metrics are listed here - https://vitess.io/docs/18.0/reference/vtorc/ui_api_metrics/.

You're right, I don't think we had the older metrics specified anywhere, but I think they can now be added to this page - https://vitess.io/docs/18.0/user-guides/configuration-advanced/reparenting/

deepthi commented on go/vt/vtctl/reparentutil/emergency_reparenter.go:
```diff
@@ -71,9 +71,20 @@ type EmergencyReparentOptions struct {
+	legacyERSSuccessCounter = stats.NewGauge("ers_success_counter", "Number of times Emergency Reparent Shard has succeeded")
+	legacyERSFailureCounter = stats.NewGauge("ers_failure_counter", "Number of times Emergency Reparent Shard has failed")
+
+	ersCounter = stats.NewCountersWithMultiLabels("emergency_reparent_shards", "Number of times Emergency Reparent Shard has been run",
```
I wonder if this can be collapsed into one stat with an additional label called result.
So that we could have something like:
```
vtctld_emergency_reparent_counts{keyspace="commerce",shard="-80",result="success"} 1
vtctld_emergency_reparent_counts{keyspace="commerce",shard="-80",result="failure"} 0
vtctld_emergency_reparent_counts{keyspace="commerce",shard="80-",result="success"} 1
vtctld_emergency_reparent_counts{keyspace="commerce",shard="80-",result="failure"} 0
```
Notice that I'm proposing dropping `shards` from the metric names and replacing it with `counts`.
I have the same suggestion for `planned_reparent` stats too.

GuptaManan100 commented:
@timvaillancourt Could you take a look at @deepthi's review?

GuptaManan100 commented:
i have addressed the review comments from Deepthi 👍

GuptaManan100 commented on go/test/endtoend/vtorc/utils/utils.go:
```diff
@@ -957,15 +959,57 @@ func WaitForSuccessfulRecoveryCount(t *testing.T, vtorcInstance *cluster.VTOrcPr
 		successfulRecoveriesMap := vars["SuccessfulRecoveries"].(map[string]interface{})
-		successCount := successfulRecoveriesMap[recoveryName]
+		successCount := int(math.Round(reflect.ValueOf(successfulRecoveriesMap[recoveryName]).Float()))
 		if successCount == countExpected {
 			return
 		}
 		time.Sleep(time.Second)
 	}
 	vars := vtorcInstance.GetVars()
 	successfulRecoveriesMap := vars["SuccessfulRecoveries"].(map[string]interface{})
-	successCount := successfulRecoveriesMap[recoveryName]
+	successCount := int(math.Round(reflect.ValueOf(successfulRecoveriesMap[recoveryName]).Float()))
```
Apparently the metric produces floats and not ints, so when we compare `successCount == countExpected` it was always false. It was only the `assert.EqualValues(t, countExpected, successCount)` which was passing. This meant that we always waited for 15 seconds whenever we entered this function!
I have fixed this too as part of this PR.
